<a href="https://colab.research.google.com/github/lovishkanther11/Deep-Learning/blob/master/Keras_Layers(ii).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Usage of activations

Activations can either be used through an Activation layer, or through the activation argument supported by all forward layers:



```
from keras.layers import Activation, Dense
model.add(Dense(64))
model.add(Activation('tanh'))
```



This is equivalent to:

```
model.add(Dense(64, activation='tanh'))
```

You can also pass an element-wise TensorFlow/Theano/CNTK function as an activation:

```
from keras import backend as K
model.add(Dense(64, activation=K.tanh))
```

#elu

keras.activations.elu(x, alpha=1.0)

Exponential linear unit.

##Arguments

1. x: Input tensor.
2. alpha: A scalar, slope of negative section.

Returns

1. The exponential linear activation: x if x > 0 and alpha * (exp(x)-1) if x < 0.

In [1]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)


# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='elu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='elu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step





Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 60000 samples, validate on 10000 samples
Epoch 1/10





60000/60000 [==============================] - 11s 188us/step - loss: 0.2916 - acc: 0.9105 - val_loss: 0.1840 - val_acc: 0.9393
Epoch 2/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.1210 - acc: 0.9624 - val_loss: 0.1447 - val_acc: 0.9555
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0819 - acc: 0.9745 - val_loss: 0.0963 - val_acc: 0.9685
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0615 - acc: 0.9804 - val_loss: 0.0846 - val_acc: 0.9747
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0507 - acc: 0.9841 - val_loss: 0.0905 - val_acc: 0.9732
Epoch 6/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.0400 - acc

#softmax

keras.activations.softmax(x, axis=-1)

##Softmax activation function.

##Arguments

1. x: Input tensor.
2. axis: Integer, axis along which the softmax normalization is applied.

##Returns

Tensor, output of softmax transformation.

##Raises

ValueError: In case dim(x) == 1.

In [2]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softmax', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softmax'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 2.2688 - acc: 0.1822 - val_loss: 2.1949 - val_acc: 0.6449
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 2.0417 - acc: 0.7224 - val_loss: 1.8538 - val_acc: 0.7615
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 1.6444 - acc: 0.7594 - val_loss: 1.4268 - val_acc: 0.7617
Epoch 4/10
60000/60000 [==============================] - 2s 28us/step - loss: 1.2334 - acc: 0.7649 - val_loss: 1.0481 - val_acc: 0.7708
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.9107 - acc: 0.7710 - val_loss: 0.7876 - val_acc: 0.7722
Epoch 6/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.7076 - acc: 0.7760 - val_loss: 0.6411 - val_acc: 0.7868
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.5993 - acc: 0.7848 - val_loss: 0.5667 - val_acc

#selu

keras.activations.selu(x)

Scaled Exponential Linear Unit (SELU).

SELU is equal to: scale * elu(x, alpha), where alpha and scale are predefined constants. The values of alpha and scale are chosen so that the mean and variance of the inputs are preserved between two consecutive layers as long as the weights are initialized correctly (see lecun_normal initialization) and the number of inputs is "large enough" (see references for more information).

##Arguments

1. x: A tensor or variable to compute the activation function for.

##Returns

The scaled exponential unit activation: scale * elu(x, alpha).

##Note

1. To be used together with the initialization "lecun_normal".
2. To be used together with the dropout variant "AlphaDropout".

In [3]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='selu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='selu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.3366 - acc: 0.8993 - val_loss: 0.1981 - val_acc: 0.9345
Epoch 2/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.1350 - acc: 0.9587 - val_loss: 0.1439 - val_acc: 0.9580
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0936 - acc: 0.9711 - val_loss: 0.1060 - val_acc: 0.9672
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0739 - acc: 0.9775 - val_loss: 0.0920 - val_acc: 0.9728
Epoch 5/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0596 - acc: 0.9815 - val_loss: 0.0980 - val_acc: 0.9745
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0488 - acc: 0.9845 - val_loss: 0.1025 - val_acc: 0.9730
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0414 - acc: 0.9872 - val_loss: 0.0826 - val_acc

#softplus

keras.activations.softplus(x)

Softplus activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softplus activation: log(exp(x) + 1).

In [4]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softplus', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softplus'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.4568 - acc: 0.8607 - val_loss: 0.1964 - val_acc: 0.9399
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.1691 - acc: 0.9480 - val_loss: 0.1531 - val_acc: 0.9540
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.1122 - acc: 0.9659 - val_loss: 0.0946 - val_acc: 0.9712
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0842 - acc: 0.9738 - val_loss: 0.0911 - val_acc: 0.9715
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0652 - acc: 0.9797 - val_loss: 0.0774 - val_acc: 0.9757
Epoch 6/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0523 - acc: 0.9838 - val_loss: 0.0874 - val_acc: 0.9751
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0437 - acc: 0.9862 - val_loss: 0.0659 - val_acc

#softsign

keras.activations.softsign(x)

##Softsign activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The softsign activation: x / (abs(x) + 1).

In [5]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='softsign', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='softsign'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.2932 - acc: 0.9119 - val_loss: 0.1728 - val_acc: 0.9472
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.1465 - acc: 0.9565 - val_loss: 0.1327 - val_acc: 0.9600
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0993 - acc: 0.9702 - val_loss: 0.0994 - val_acc: 0.9703
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0750 - acc: 0.9769 - val_loss: 0.0937 - val_acc: 0.9719
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0584 - acc: 0.9823 - val_loss: 0.0836 - val_acc: 0.9746
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0449 - acc: 0.9861 - val_loss: 0.0756 - val_acc: 0.9752
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0358 - acc: 0.9889 - val_loss: 0.0715 - val_acc

#relu

keras.activations.relu(x, alpha=0.0, max_value=None, threshold=0.0)

##Rectified Linear Unit.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = alpha * (x - threshold) otherwise.

##Arguments

1. x: Input tensor.
2. alpha: float. Slope of the negative part. Defaults to zero.
3. max_value: float. Saturation threshold.
4. threshold: float. Threshold value for thresholded activation.

##Returns

A tensor.

In [6]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='relu', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.2258 - acc: 0.9316 - val_loss: 0.2088 - val_acc: 0.9345
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0850 - acc: 0.9745 - val_loss: 0.1019 - val_acc: 0.9685
Epoch 3/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.0563 - acc: 0.9824 - val_loss: 0.0681 - val_acc: 0.9811
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0407 - acc: 0.9873 - val_loss: 0.0732 - val_acc: 0.9784
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0302 - acc: 0.9904 - val_loss: 0.0786 - val_acc: 0.9805
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0231 - acc: 0.9931 - val_loss: 0.0857 - val_acc: 0.9799
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0182 - acc: 0.9944 - val_loss: 0.0912 - val_acc

#tanh

keras.activations.tanh(x)

##Hyperbolic tangent activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The hyperbolic activation: tanh(x) = (exp(x) - exp(-x)) / (exp(x) + exp(-x))

In [7]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='tanh', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='tanh'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 35us/step - loss: 0.2856 - acc: 0.9144 - val_loss: 0.1543 - val_acc: 0.9541
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1225 - acc: 0.9634 - val_loss: 0.1180 - val_acc: 0.9636
Epoch 3/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0802 - acc: 0.9752 - val_loss: 0.0848 - val_acc: 0.9723
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0567 - acc: 0.9824 - val_loss: 0.0762 - val_acc: 0.9774
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0420 - acc: 0.9872 - val_loss: 0.0704 - val_acc: 0.9768
Epoch 6/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0314 - acc: 0.9907 - val_loss: 0.0657 - val_acc: 0.9784
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0222 - acc: 0.9930 - val_loss: 0.0649 - val_acc

#sigmoid

keras.activations.sigmoid(x)

##Sigmoid activation function.

##Arguments

1. x: Input tensor.

##Returns

1. The sigmoid activation: 1 / (1 + exp(-x)).

In [8]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.4889 - acc: 0.8592 - val_loss: 0.2569 - val_acc: 0.9202
Epoch 2/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.2189 - acc: 0.9341 - val_loss: 0.1808 - val_acc: 0.9453
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.1571 - acc: 0.9524 - val_loss: 0.1357 - val_acc: 0.9586
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1186 - acc: 0.9649 - val_loss: 0.1322 - val_acc: 0.9602
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0935 - acc: 0.9723 - val_loss: 0.0973 - val_acc: 0.9694
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0743 - acc: 0.9774 - val_loss: 0.0862 - val_acc: 0.9730
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0622 - acc: 0.9807 - val_loss: 0.0852 - val_acc

#hard_sigmoid

keras.activations.hard_sigmoid(x)

##Hard sigmoid activation function.

Faster to compute than sigmoid activation.

##Arguments

1. x: Input tensor.

##Returns

1. Hard sigmoid activation:

```
0 if x < -2.5
1 if x > 2.5
0.2 * x + 0.5 if -2.5 <= x <= 2.5.
```

In [9]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='hard_sigmoid', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='hard_sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.5028 - acc: 0.8541 - val_loss: 0.2881 - val_acc: 0.9111
Epoch 2/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.2146 - acc: 0.9359 - val_loss: 0.1719 - val_acc: 0.9460
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1503 - acc: 0.9547 - val_loss: 0.1304 - val_acc: 0.9581
Epoch 4/10
60000/60000 [==============================] - 2s 34us/step - loss: 0.1128 - acc: 0.9661 - val_loss: 0.1235 - val_acc: 0.9604
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0877 - acc: 0.9733 - val_loss: 0.1021 - val_acc: 0.9702
Epoch 6/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0708 - acc: 0.9788 - val_loss: 0.0822 - val_acc: 0.9746
Epoch 7/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0574 - acc: 0.9819 - val_loss: 0.0748 - val_acc

#exponential

keras.activations.exponential(x)

Exponential (base e) activation function.

##Arguments

1. x: Input tensor.

##Returns

1. Exponential activation: exp(x).

In [10]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='exponential', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='exponential'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 14.5062 - acc: 0.0988 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 4/10
60000/60000 [==============================] - 2s 31us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 6/10
60000/60000 [==============================] - 2s 30us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.5740 - val_acc: 0.0958
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 14.5283 - acc: 0.0986 - val_loss: 14.

#linear

keras.activations.linear(x)

Linear (i.e. identity) activation function.

##Arguments

1. x: Input tensor.

##Returns

Input tensor, unchanged.

In [11]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation='linear', input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation='linear'))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.4217 - acc: 0.8767 - val_loss: 0.3212 - val_acc: 0.9105
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.3447 - acc: 0.9016 - val_loss: 0.3398 - val_acc: 0.9068
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.3211 - acc: 0.9098 - val_loss: 0.3149 - val_acc: 0.9095
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.3095 - acc: 0.9124 - val_loss: 0.3146 - val_acc: 0.9097
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.2994 - acc: 0.9163 - val_loss: 0.2997 - val_acc: 0.9142
Epoch 6/10
60000/60000 [==============================] - 2s 27us/step - loss: 0.2937 - acc: 0.9173 - val_loss: 0.3186 - val_acc: 0.9097
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.2896 - acc: 0.9191 - val_loss: 0.2924 - val_acc

#LeakyReLU

keras.layers.LeakyReLU(alpha=0.3)

Leaky version of a Rectified Linear Unit.

It allows a small gradient when the unit is not active: 
```
f(x) = alpha * x for x < 0
f(x) = x for x >= 0.
```

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

alpha: float >= 0. Negative slope coefficient.

In [12]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.LeakyReLU(alpha=0.3), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.LeakyReLU(alpha=0.3)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as LeakyReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.2653 - acc: 0.9190 - val_loss: 0.1671 - val_acc: 0.9516
Epoch 2/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.1260 - acc: 0.9620 - val_loss: 0.1234 - val_acc: 0.9624
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0935 - acc: 0.9714 - val_loss: 0.1144 - val_acc: 0.9639
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0747 - acc: 0.9771 - val_loss: 0.0886 - val_acc: 0.9730
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0633 - acc: 0.9801 - val_loss: 0.1082 - val_acc: 0.9691
Epoch 6/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0548 - acc: 0.9829 - val_loss: 0.0844 - val_acc: 0.9772
Epoch 7/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0481 - acc: 0.9848 - val_loss: 0.0921 - val_acc

#PReLU

keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)

Parametric Rectified Linear Unit.

It follows: f(x) = alpha * x for x < 0, f(x) = x for x >= 0, where alpha is a learned array with the same shape as x.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha_initializer: initializer function for the weights.
2. alpha_regularizer: regularizer for the weights.
3. alpha_constraint: constraint for the weights.
4. shared_axes: the axes along which to share learnable parameters for the activation function. For example, if the incoming feature maps are from a 2D convolution with output shape (batch, height, width, channels), and you wish to share parameters across space so that each filter only has one set of parameters, set shared_axes=[1, 2].

In [13]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.PReLU(alpha_initializer='zeros', alpha_regularizer=None, alpha_constraint=None, shared_axes=None)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as PReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 2s 41us/step - loss: 0.2271 - acc: 0.9310 - val_loss: 0.1070 - val_acc: 0.9677
Epoch 2/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0841 - acc: 0.9736 - val_loss: 0.0937 - val_acc: 0.9702
Epoch 3/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0551 - acc: 0.9827 - val_loss: 0.0708 - val_acc: 0.9795
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0390 - acc: 0.9880 - val_loss: 0.0751 - val_acc: 0.9799
Epoch 5/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0303 - acc: 0.9904 - val_loss: 0.0711 - val_acc: 0.9813
Epoch 6/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0227 - acc: 0.9931 - val_loss: 0.0773 - val_acc: 0.9818
Epoch 7/10
60000/60000 [==============================] - 2s 33us/step - loss: 0.0178 - acc: 0.9942 - val_loss: 0.1027 - val_acc

#ELU

keras.layers.ELU(alpha=1.0)

Exponential Linear Unit.

It follows: f(x) =  alpha * (exp(x) - 1.) for x < 0, f(x) = x for x >= 0.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. alpha: scale for the negative factor.

In [14]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ELU(alpha=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ELU(alpha=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ELU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 44us/step - loss: 0.2923 - acc: 0.9123 - val_loss: 0.1674 - val_acc: 0.9480
Epoch 2/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.1225 - acc: 0.9626 - val_loss: 0.1034 - val_acc: 0.9696
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0841 - acc: 0.9743 - val_loss: 0.0897 - val_acc: 0.9726
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0639 - acc: 0.9795 - val_loss: 0.0844 - val_acc: 0.9741
Epoch 5/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.0506 - acc: 0.9839 - val_loss: 0.0728 - val_acc: 0.9780
Epoch 6/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0402 - acc: 0.9870 - val_loss: 0.0839 - val_acc: 0.9751
Epoch 7/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0337 - acc: 0.9893 - val_loss: 0.0825 - val_acc

#ThresholdedReLU

keras.layers.ThresholdedReLU(theta=1.0)

Thresholded Rectified Linear Unit.

It follows: f(x) = x for x > theta, f(x) = 0 otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. theta: float >= 0. Threshold location of activation.

In [15]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ThresholdedReLU(theta=1.0), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ThresholdedReLU(theta=1.0)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ThresholdedReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 42us/step - loss: 0.7112 - acc: 0.7548 - val_loss: 0.2011 - val_acc: 0.9372
Epoch 2/10
60000/60000 [==============================] - 2s 32us/step - loss: 0.1523 - acc: 0.9550 - val_loss: 0.1139 - val_acc: 0.9663
Epoch 3/10
60000/60000 [==============================] - 2s 28us/step - loss: 0.1034 - acc: 0.9684 - val_loss: 0.1038 - val_acc: 0.9675
Epoch 4/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0773 - acc: 0.9766 - val_loss: 0.0967 - val_acc: 0.9708
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0613 - acc: 0.9809 - val_loss: 0.0866 - val_acc: 0.9744
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0506 - acc: 0.9842 - val_loss: 0.0858 - val_acc: 0.9755
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0421 - acc: 0.9868 - val_loss: 0.0884 - val_acc

#Softmax

keras.layers.Softmax(axis=-1)

Softmax activation function.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. axis: Integer, axis along which the softmax normalization is applied.

In [16]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.Softmax(axis=-1), input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.Softmax(axis=-1)))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as Softmax) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 45us/step - loss: 2.2691 - acc: 0.2051 - val_loss: 2.1956 - val_acc: 0.7002
Epoch 2/10
60000/60000 [==============================] - 2s 32us/step - loss: 2.0398 - acc: 0.7422 - val_loss: 1.8486 - val_acc: 0.7576
Epoch 3/10
60000/60000 [==============================] - 2s 30us/step - loss: 1.6349 - acc: 0.7556 - val_loss: 1.4129 - val_acc: 0.7599
Epoch 4/10
60000/60000 [==============================] - 2s 32us/step - loss: 1.2162 - acc: 0.7667 - val_loss: 1.0281 - val_acc: 0.7611
Epoch 5/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.8912 - acc: 0.7707 - val_loss: 0.7706 - val_acc: 0.7713
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.6944 - acc: 0.7752 - val_loss: 0.6301 - val_acc: 0.7739
Epoch 7/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.5927 - acc: 0.7799 - val_loss: 0.5659 - val_acc

#ReLU

keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)

Rectified Linear Unit activation function.

With default values, it returns element-wise max(x, 0).

Otherwise, it follows: f(x) = max_value for x >= max_value, f(x) = x for threshold <= x < max_value, f(x) = negative_slope * (x - threshold) otherwise.

##Input shape

Arbitrary. Use the keyword argument input_shape (tuple of integers, does not include the samples axis) when using this layer as the first layer in a model.

##Output shape

Same shape as the input.

##Arguments

1. max_value: float >= 0. Maximum activation value.
2. negative_slope: float >= 0. Negative slope coefficient.
3. threshold: float. Threshold value for thresholded activation.

In [17]:
# Build neural network
model = models.Sequential()
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 3s 44us/step - loss: 0.2259 - acc: 0.9311 - val_loss: 0.1087 - val_acc: 0.9676
Epoch 2/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0849 - acc: 0.9734 - val_loss: 0.1033 - val_acc: 0.9694
Epoch 3/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0558 - acc: 0.9821 - val_loss: 0.0814 - val_acc: 0.9750
Epoch 4/10
60000/60000 [==============================] - 2s 29us/step - loss: 0.0401 - acc: 0.9875 - val_loss: 0.0900 - val_acc: 0.9745
Epoch 5/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0298 - acc: 0.9908 - val_loss: 0.0774 - val_acc: 0.9801
Epoch 6/10
60000/60000 [==============================] - 2s 31us/step - loss: 0.0218 - acc: 0.9932 - val_loss: 0.0817 - val_acc: 0.9807
Epoch 7/10
60000/60000 [==============================] - 2s 30us/step - loss: 0.0178 - acc: 0.9943 - val_loss: 0.0838 - val_acc



1.   **elu**:
      acc: 0.9938 - val_loss: 0.0883 - val_acc: 0.9796
2.   **softmax**:
      acc: 0.8702 - val_loss: 0.3958 - val_acc: 0.8675
3.   **selu**:
      acc: 0.9914 - val_loss: 0.0934 - val_acc: 0.9791
4.   **softplus**:
      acc: 0.9921 - val_loss: 0.0859 - val_acc: 0.9782
5.   **softsign**:
      acc: 0.9951 - val_loss: 0.0741 - val_acc: 0.9773
6.   **relu**:
      acc: 0.9968 - val_loss: 0.1328 - val_acc: 0.9753
7.   **tanh**:
      acc: 0.9976 - val_loss: 0.0736 - val_acc: 0.9799
8.   **sigmoid**:
      acc: 0.9886 - val_loss: 0.0678 - val_acc: 0.9796
9.   **hard_sigmoid**:
      acc: 0.9909 - val_loss: 0.0785 - val_acc: 0.9765
10.  **exponential**:
      acc: 0.1124 - val_loss: 14.2887 - val_acc: 0.1135
11.  **linear**:
      acc: 0.9231 - val_loss: 0.2988 - val_acc: 0.9158
12.  **LeakyReLu**:
      acc: 0.9887 - val_loss: 0.0924 - val_acc: 0.9778
13.  **PReLu**:
      acc: 0.9968 - val_loss: 0.0903 - val_acc: 0.9820
14.  **ELU**:
      acc: 0.9935 - val_loss: 0.0861 - val_acc: 0.9795
15.  **ThresholdedReLu**:
      acc: 0.1124 - val_loss: 2.3010 - val_acc: 0.1135
16.  **Softmax**:
      acc: 0.8710 - val_loss: 0.3985 - val_acc: 0.8658
17.  **ReLu**:
      acc: 0.9968 - val_loss: 0.0937 - val_acc: 0.9829

###The best activation function is ReLu which gives an accuracy of 0.9968 and a validation accuracy of 0.9829.




##Fashion MNIST

In [18]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# reshape to be [samples][width][height][channels]
X_train = X_train.reshape((X_train.shape[0], 28, 28, 1))
X_test = X_test.reshape((X_test.shape[0], 28, 28, 1))

# Reshape data
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

# Categorically encode labels
y_train = to_categorical(y_train, NUM_CLASSES)
y_test = to_categorical(y_test, NUM_CLASSES)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)


# Build neural network
model = models.Sequential()
model.add(Conv2D(512, (28, 28), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:]))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(NUM_ROWS * NUM_COLS,)))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
))
model.add(Dense(10, activation='softmax'))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, y_train,
          batch_size=BATCH_SIZE,
          epochs=EPOCHS,
          verbose=1,
          validation_data=(X_test, y_test))

score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

4423680/4422102 [==============================] - 1s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 8s 141us/step - loss: 0.5145 - acc: 0.8153 - val_loss: 0.4247 - val_acc: 0.8514
Epoch 2/10
60000/60000 [==============================] - 2s 36us/step - loss: 0.3738 - acc: 0.8642 - val_loss: 0.3773 - val_acc: 0.8637
Epoch 3/10
60000/60000 [==============================] - 2s 38us/step - loss: 0.3331 - acc: 0.8771 - val_loss: 0.3877 - val_acc: 0.8571
Epoch 4/10
60000/60000 [==============================] - 2s 37us/step - loss: 0.3083 - acc: 0.8851 - val_loss: 0.3577 - val_acc: 0.8675
Epoch 5/10
60000/60000 [==============================] - 2s 39us/step - loss: 0.2897 - acc: 0.8928 - val_loss: 0.3431 - val_acc: 0.8721
Epoch 6/10
60000/60000 [==============================] - 2s 39us/step - loss: 0.2719 - acc: 0.8979 - val_loss: 0.3244 - val_acc: 0.8803
Epoch 7/10
60000/60000 [==============================] - 2s 40us/step - loss: 0.2560 - acc: 0.9037 - val_loss: 0.3373 - val_ac

##CIFAR 10

In [19]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar10
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

#Model building
initializer = keras.initializers.lecun_normal(seed=None)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:], kernel_initializer=initializer, bias_initializer=initializer))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Dense(10, activation='softmax', kernel_initializer=initializer, bias_initializer=initializer))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

170500096/170498071 [==============================] - 2s 0us/step



/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 5s 99us/step - loss: 1.9555 - acc: 0.2987 - val_loss: 1.7500 - val_acc: 0.3701
Epoch 2/10
50000/50000 [==============================] - 3s 70us/step - loss: 1.6967 - acc: 0.3929 - val_loss: 1.6457 - val_acc: 0.4073
Epoch 3/10
50000/50000 [==============================] - 4s 71us/step - loss: 1.5995 - acc: 0.4280 - val_loss: 1.5810 - val_acc: 0.4348
Epoch 4/10
50000/50000 [==============================] - 4s 71us/step - loss: 1.5281 - acc: 0.4547 - val_loss: 1.4908 - val_acc: 0.4653
Epoch 5/10
50000/50000 [==============================] - 4s 72us/step - loss: 1.4745 - acc: 0.4724 - val_loss: 1.5047 - val_acc: 0.4707
Epoch 6/10
50000/50000 [==============================] - 4s 71us/step - loss: 1.4299 - acc: 0.4899 - val_loss: 1.4139 - val_acc: 0.5000
Epoch 7/10
50000/50000 [==============================] - 4s 71us/step - loss: 1.3881 - acc: 0.5050 - val_loss: 1.4453 - val_acc

##CIFAR 100

In [20]:
import keras
from keras import models
from keras.layers import Dense, Dropout, Conv2D,Flatten
from keras.utils import to_categorical
from keras.datasets import cifar100
from keras.utils.vis_utils import model_to_dot
from IPython.display import SVG
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers

#import dataset
(X_train, y_train), (X_test, y_test) = cifar100.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000,32,32,3)
X_test = X_test.reshape(10000,32,32,3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 100)
Y_test =  to_categorical(y_test, 100)

adamax=keras.optimizers.Adamax(lr=0.002, beta_1=0.9, beta_2=0.999)

#Model building
initializer = keras.initializers.lecun_normal(seed=None)


# Build neural network
model = models.Sequential()
model.add(Conv2D(1024, (32, 32), padding="valid", activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0), input_shape=X_train.shape[1:],kernel_initializer=initializer, bias_initializer=initializer))
model.add(Flatten())
model.add(Dense(512, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
, input_shape=(32 * 32,),kernel_initializer=initializer, bias_initializer=initializer))
model.add(Dense(256, activation=keras.layers.ReLU(max_value=None, negative_slope=0.0, threshold=0.0)
,kernel_initializer=initializer, bias_initializer=initializer))
model.add(Dense(100, activation='softmax',kernel_initializer=initializer, bias_initializer=initializer))

# Compile model
model.compile(optimizer='adamax',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train model
model.fit(X_train, Y_train,
          batch_size=128,
          epochs=10,
          verbose=1,
          validation_data=(X_test, Y_test))

score = model.evaluate(X_test, Y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

169009152/169001437 [==============================] - 2s 0us/step


/usr/local/lib/python3.6/dist-packages/keras/activations.py:235: UserWarning: Do not pass a layer instance (such as ReLU) as the activation argument of another layer. Instead, advanced activation layers should be used just like any other layer in a model.
  identifier=identifier.__class__.__name__))


Train on 50000 samples, validate on 10000 samples
Epoch 1/10
50000/50000 [==============================] - 5s 94us/step - loss: 4.3176 - acc: 0.0457 - val_loss: 4.0024 - val_acc: 0.0749
Epoch 2/10
50000/50000 [==============================] - 4s 71us/step - loss: 3.8422 - acc: 0.1107 - val_loss: 3.7245 - val_acc: 0.1342
Epoch 3/10
50000/50000 [==============================] - 3s 70us/step - loss: 3.6262 - acc: 0.1466 - val_loss: 3.5985 - val_acc: 0.1533
Epoch 4/10
50000/50000 [==============================] - 4s 72us/step - loss: 3.4898 - acc: 0.1718 - val_loss: 3.4938 - val_acc: 0.1734
Epoch 5/10
50000/50000 [==============================] - 4s 71us/step - loss: 3.3669 - acc: 0.1930 - val_loss: 3.4541 - val_acc: 0.1817
Epoch 6/10
50000/50000 [==============================] - 4s 71us/step - loss: 3.2701 - acc: 0.2123 - val_loss: 3.3383 - val_acc: 0.2025
Epoch 7/10
50000/50000 [==============================] - 4s 71us/step - loss: 3.1841 - acc: 0.2259 - val_loss: 3.3148 - val_acc

#Thank you for completing this notebook